File: packages/core/src/privacy/PrivacyGuard.ts
import { Redaction, RedactionResult, RedactionType } from "./Redaction";
import { PrivacyPolicyConfig, defaultPrivacyPolicy } from "./PrivacyPolicy";

const PATTERNS: Record<RedactionType, RegExp> = {
  email: /[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2}/g,
  phone: /(\+?1?\s?)?(\(?\d{3}\)?[\s.-]?)?\d{3}[\s.-]?\d{4}/g,
  token: /[A-Za-z0-9_-]{20}/g,
  api_key: /(sk|pk)_[A-Za-z0-9_-]{20}/g,
  credit_card: /\d{4}[\s-]?\d{4}[\s-]?\d{4}[\s-]?\d{4}/g,
  ssn: /\d{3}-\d{2}-\d{4}/g,
};

const REPLACEMENTS: Record<RedactionType, string> = {
  email: "[REDACTED_EMAIL]",
  phone: "[REDACTED_PHONE]",
  token: "[REDACTED_TOKEN]",
  api_key: "[REDACTED_API_KEY]",
  credit_card: "[REDACTED_CREDIT_CARD]",
  ssn: "[REDACTED_SSN]",
};

export class PrivacyGuard {
  private policy: PrivacyPolicyConfig;
  private redactionLog: Redaction[];

  constructor(policy?: Partial<PrivacyPolicyConfig>) {
    this.policy = { ...defaultPrivacyPolicy, ...policy };
    this.redactionLog = [];
  }

  redact(text: string): RedactionResult {
    let redactedText = text;
    const redactions: Redaction[] = [];

    for (const type of this.policy.enabledRedactions) {
      const pattern = PATTERNS[type];
      const replacement = REPLACEMENTS[type];

      const matches = Array.from(text.matchAll(pattern));

      for (const match of matches) {
        if (match.index === undefined) continue;

        const redaction: Redaction = {
          type,
          original: match[0],
          replacement,
          position: {
            start: match.index,
            end: match.index + match[0].length,
          },
        };

        redactions.push(redaction);
        redactedText = redactedText.replace(match[0], replacement);
      }
    }

    if (this.policy.logRedactions) {
      this.redactionLog.push(...redactions);
    }

    return {
      redactedText,
      redactions,
      safe: redactions.length === 0,
    };
  }

  getRedactionLog(): Redaction[] {
    return [...this.redactionLog];
  }

  clearRedactionLog(): void {
    this.redactionLog = [];
  }
}

export function createDefaultPrivacyGuard(): PrivacyGuard {
  return new PrivacyGuard();
}

File: packages/compose/src/main/kotlin/ai/mobileagentos/compose/privacy/PrivacyGuard.kt
package ai.mobileagentos.compose.privacy

class PrivacyGuard(
    policy: PrivacyPolicyConfig = defaultPrivacyPolicy
) {
    private val policy: PrivacyPolicyConfig = policy
    private val redactionLog: MutableList<Redaction> = mutableListOf()
    
    private val patterns = mapOf(
        RedactionType.EMAIL to Regex("[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}"),
        RedactionType.PHONE to Regex("(\\+?1?\\s?)?(\\(?\\d{3}\\)?[\\s.-]?)?\\d{3}[\\s.-]?\\d{4}"),
        RedactionType.TOKEN to Regex("[A-Za-z0-9_-]{20,}"),
        RedactionType.API_KEY to Regex("(sk|pk)_[A-Za-z0-9_-]{20,}"),
        RedactionType.CREDIT_CARD to Regex("\\d{4}[\\s-]?\\d{4}[\\s-]?\\d{4}[\\s-]?\\d{4}"),
        RedactionType.SSN to Regex("\\d{3}-\\d{2}-\\d{4}")
    )
    
    private val replacements = mapOf(
        RedactionType.EMAIL to "[REDACTED_EMAIL]",
        RedactionType.PHONE to "[REDACTED_PHONE]",
        RedactionType.TOKEN to "[REDACTED_TOKEN]",
        RedactionType.API_KEY to "[REDACTED_API_KEY]",
        RedactionType.CREDIT_CARD to "[REDACTED_CREDIT_CARD]",
        RedactionType.SSN to "[REDACTED_SSN]"
    )
    
    fun redact(text: String): PrivacyResult {
        var redactedText = text
        val redactions = mutableListOf<Redaction>()
        
        for (type in policy.enabledRedactions) {
            val pattern = patterns[type] ?: continue
            val replacement = replacements[type] ?: continue
            
            pattern.findAll(text).forEach { match ->
                val redaction = Redaction(
                    type = type,
                    original = match.value,
                    replacement = replacement,
                    position = match.range
                )
                redactions.add(redaction)
                redactedText = redactedText.replace(match.value, replacement)
            }
        }
        
        if (policy.logRedactions) {
            redactionLog.addAll(redactions)
        }
        
        return PrivacyResult(
            redactedText = redactedText,
            redactions = redactions,
            safe = redactions.isEmpty()
        )
    }
    
    fun getRedactionLog(): List<Redaction> = redactionLog.toList()
    
    fun clearRedactionLog() {
        redactionLog.clear()
    }
}